### PR TITLE
Improved Eval function

### PR DESCRIPTION
**Pull request by MTGOStark**
This is an eval curve determined and normalized by MTGOStark and syjytg, then fine-tuned by AiledlMorn to fit the tan function. Based on SF positions and conventional evaluations. It should scale well through all positions and phases of the game. A preview of the curve (and similar curves) can be found here: https://images-ext-2.discordapp.net/external/B3OEYfVT-bF4jn1su5vrIJyo5Op8HSJ1EdWxYwYiTKc/https/i.imgur.com/h0ZFkxf.png

### CODE REVIEW

**Comment by Akababa** [comment]
I obtained some stockfish evals from removing pieces from the starting position and ran a couple regressions here: https://docs.google.com/spreadsheets/d/16SqOWzrht3ZHsli0jntUrsXY5XKSCSJiUrvtwvRjLrs/edit#gid=0

Higher variance and larger sample size would be nice but it tracks the stockfish eval consistently (albeit lower). Another possible formula is atanh(2x-1)*9.72.

**Comment by man4** [comment]
Note that 972\*atanh(2x-1) is the same as 486\*ln(x/(1-x)), which is just the old logit function with a different constant: http://www.wolframalpha.com/input/?i=486log(x%2F(1-x))+vs+972atanh(2x-1)+from+0+to+1

**Comment by syjytg** [comment]
I suggest 290.68tan(3.096(x-0.5)). It is very similar to 300tan(3(x-0.5)) up to 90+% but 290.68tan(3.096(x-0.5)) hits end point of +127.49 (12749cp) which is very close to SF mate score.

**Comment by killerducky** [comment]
@MTGOStark @syjytg can you guys decide among each other on these, pick one, and close the other?

**Comment by syjytg** [comment]
I say we take 290.680623072 * tan(3.096181612 * (feval - 0.5)) because the 2 graphs are extremely very similar up to 90+%. However the endpoint of 290.680623072 * tan(3.096181612 * (feval - 0.5)) is exactly +128 (12800cp) which is SF mate score.

**Comment by dubslow** [comment]
First, I don't like changing the multiplier inside `tan` away from pi, for mostly aesthetic reasons. It's nice to say a won game is infinitely better.

Secondly, I think the atanh approach is also more aesthetic, it far more closely represents what's going on inside the neurons, so it's almost like undoing the neuron stuff to get a cp evaluation. Tweak it sure, but real exponentials are more appropriate in my book than imaginary exponentials, not unlike [the original comment](https://github.com/glinscott/leela-chess/commit/91fe0bcf03989843cbc9231b530a1c6b2f4e3d42#diff-70aab7683fe64f9428b8bc47d1d183c0L246) near the conversion code.

**Comment by man4** [comment]
The old approach was equivalent to using atanh, since atanh is just a rescaled logit function. However, it was replaced in favor of tan, because the tangent function gives higher evals for winrates near 100%.

The atanh/logit function seems like the more "elegant" solution mathematically. The issue is that in practice, it doesn't differentiate greatly between mate positions and positions where you're a queen up. 

One possible solution is to output the winrate of only the subtree corresponding to the root move that is actually played, instead of a weighted average over all root moves. This makes sense intuitively, because the winrates of the root moves you chose _not_ to play shouldn't matter. In near-mate (especially mate in 1) positions, this would help bring the winrate much closer to 100%, and atanh/logit would transform the winrate to a high eval. One small detail about this approach is that often the winrate will be _exactly_ 100% in a mate in 1 position, causing the logit to be undefined, so it's perhaps best to use add-one smoothing when outputting the winrate.

**Comment by dubslow** [comment]
Well if we're talking about extending the tan domain effectively beyond the inputs, to have a finite mate value, why not do the same thing for arctan as well?

**Comment by MTGOStark** [comment]
@killerducky the newer one from PR#329 looks good; it's this one with endpoints tweaked to correspond with SF mate score.

**Comment by glinscott** [comment]
Thanks!
